File: apps/desktop/src/lib/history.ts
import { invoke } from "@tauri-apps/api/core";

export type AccuracyFlag =
  | "accurate"
  | "inaccurate_raw"
  | "inaccurate_polished"
  | "unknown";

export type HistoryActionKind =
  | "copy"
  | "reinsert"
  | "export"
  | "save_draft"
  | "clipboard_backup";

export type HistoryPostAction = {
  kind: HistoryActionKind;
  timestampMs: number;
  detail: Record<string, unknown>;
};

export type HistoryEntry = {
  sessionId: string;
  startedAtMs: number;
  completedAtMs: number;
  durationMs: number;
  locale?: string | null;
  appIdentifier?: string | null;
  appVersion?: string | null;
  confidenceScore?: number | null;
  rawTranscript: string;
  polishedTranscript: string;
  preview: string;
  accuracyFlag: AccuracyFlag;
  accuracyRemarks?: string | null;
  postActions: HistoryPostAction[];
  metadata: Record<string, unknown>;
};

export type HistoryPage = {
  entries: HistoryEntry[];
  nextOffset: number | null;
  total: number | null;
};

export type HistoryQuery = {
  keyword?: string;
  locale?: string;
  appIdentifier?: string;
  limit?: number;
  offset?: number;
};

export type HistoryAccuracyRequest = {
  sessionId: string;
  flag: AccuracyFlag;
  remarks?: string;
};

export type HistoryActionRequest = {
  sessionId: string;
  action: HistoryActionKind;
  detail?: Record<string, unknown>;
};

const entryCache = new Map<string, HistoryEntry>();
const pageCache = new Map<string, HistoryPage>();

function invalidateSessionCache(sessionId: string) {
  entryCache.delete(sessionId);
  for (const [key, page] of pageCache.entries()) {
    if (page.entries.some((entry) => entry.sessionId === sessionId)) {
      pageCache.delete(key);
    }
  }
}

function cacheKey(query: HistoryQuery): string {
  const normalized: HistoryQuery = {
    keyword: query.keyword?.trim() || undefined,
    locale: query.locale?.trim() || undefined,
    appIdentifier: query.appIdentifier?.trim() || undefined,
    limit: query.limit ?? undefined,
    offset: query.offset ?? undefined,
  };
  return JSON.stringify(normalized);
}

function clone<T>(value: T): T {
  return structuredClone ? structuredClone(value) : JSON.parse(JSON.stringify(value));
}

export function clearHistoryCache() {
  entryCache.clear();
  pageCache.clear();
}

export async function searchHistory(query: HistoryQuery = {}): Promise<HistoryPage> {
  const key = cacheKey(query);
  if (pageCache.has(key)) {
    return clone(pageCache.get(key)!);
  }

  const page = await invoke<HistoryPage>("session_history_search", { query });
  page.entries.forEach((entry) => entryCache.set(entry.sessionId, entry));
  pageCache.set(key, page);
  return clone(page);
}

export async function loadHistoryEntry(sessionId: string): Promise<HistoryEntry | null> {
  if (entryCache.has(sessionId)) {
    return clone(entryCache.get(sessionId)!);
  }

  const entry = await invoke<HistoryEntry | null>("session_history_entry", { sessionId });
  if (entry) {
    entryCache.set(sessionId, entry);
    return clone(entry);
  }
  return null;
}

export async function markHistoryAccuracy(request: HistoryAccuracyRequest): Promise<void> {
  await invoke<void>("session_history_mark_accuracy", {
    update: request,
  });
  invalidateSessionCache(request.sessionId);
}

export async function recordHistoryAction(
  request: HistoryActionRequest,
): Promise<HistoryPostAction[]> {
  const actions = await invoke<HistoryPostAction[]>("session_history_append_action", {
    request,
  });
  invalidateSessionCache(request.sessionId);
  return clone(actions);
}

File: core/src/session/history.rs
//! Session history domain models and DTOs for persistence and IPC layers.

use serde::{Deserialize, Serialize};
use serde_json::json;
use std::cmp::min;

/// History retention in hours. Sessions older than this window will be purged.
pub const HISTORY_RETENTION_HOURS: i64 = 48;
/// Retention window expressed in milliseconds.
pub const HISTORY_RETENTION_MS: i64 = HISTORY_RETENTION_HOURS * 60 * 60 * 1_000;
/// Preview length surfaced in UI search results.
pub const HISTORY_PREVIEW_LIMIT: usize = 120;

/// Accuracy flag captured from user feedback flows.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AccuracyFlag {
    Accurate,
    InaccurateRaw,
    InaccuratePolished,
    Unknown,
}

impl Default for AccuracyFlag {
    fn default() -> Self {
        AccuracyFlag::Unknown
    }
}

impl AccuracyFlag {
    /// Returns the canonical string value persisted in the database.
    pub fn as_str(&self) -> &'static str {
        match self {
            AccuracyFlag::Accurate => "accurate",
            AccuracyFlag::InaccurateRaw => "inaccurate_raw",
            AccuracyFlag::InaccuratePolished => "inaccurate_polished",
            AccuracyFlag::Unknown => "unknown",
        }
    }

    pub fn from_db(value: Option<&str>) -> Self {
        match value {
            Some("accurate") => AccuracyFlag::Accurate,
            Some("inaccurate_raw") => AccuracyFlag::InaccurateRaw,
            Some("inaccurate_polished") => AccuracyFlag::InaccuratePolished,
            _ => AccuracyFlag::Unknown,
        }
    }
}

/// Post actions triggered from history detail (copy, reinsert, export, etc.).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum HistoryActionKind {
    Copy,
    Reinsert,
    Export,
    SaveDraft,
    ClipboardBackup,
}

impl HistoryActionKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            HistoryActionKind::Copy => "copy",
            HistoryActionKind::Reinsert => "reinsert",
            HistoryActionKind::Export => "export",
            HistoryActionKind::SaveDraft => "save_draft",
            HistoryActionKind::ClipboardBackup => "clipboard_backup",
        }
    }
}

/// Metadata describing a user action taken on a history entry.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct HistoryPostAction {
    pub kind: HistoryActionKind,
    pub timestamp_ms: i64,
    #[serde(default = "HistoryPostAction::default_detail")]
    pub detail: serde_json::Value,
}

impl HistoryPostAction {
    fn default_detail() -> serde_json::Value {
        json!({})
    }

    pub fn clipboard_backup(timestamp_ms: i64) -> Self {
        Self {
            kind: HistoryActionKind::ClipboardBackup,
            timestamp_ms,
            detail: Self::default_detail(),
        }
    }
}

/// Snapshot of a completed session ready for persistence.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SessionSnapshot {
    pub session_id: String,
    pub started_at_ms: i64,
    pub completed_at_ms: i64,
    #[serde(default)]
    pub locale: Option<String>,
    #[serde(default)]
    pub app_identifier: Option<String>,
    #[serde(default)]
    pub app_version: Option<String>,
    #[serde(default)]
    pub confidence_score: Option<f32>,
    pub raw_transcript: String,
    pub polished_transcript: String,
    #[serde(default)]
    pub metadata: serde_json::Value,
    #[serde(default)]
    pub post_actions: Vec<HistoryPostAction>,
}

impl SessionSnapshot {
    pub fn duration_ms(&self) -> i64 {
        (self.completed_at_ms - self.started_at_ms).max(0)
    }

    pub fn expires_at_ms(&self) -> i64 {
        self.completed_at_ms + HISTORY_RETENTION_MS
    }

    pub fn preview(&self) -> String {
        let mut text = if !self.polished_transcript.trim().is_empty() {
            self.polished_transcript.clone()
        } else {
            self.raw_transcript.clone()
        };

        if text.len() > HISTORY_PREVIEW_LIMIT {
            let preview_end = min(text.len(), HISTORY_PREVIEW_LIMIT);
            text.truncate(preview_end);
            text.push_str("…");
        }

        text
    }
}

/// Query filters used when listing history entries.
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct HistoryQuery {
    #[serde(default)]
    pub keyword: Option<String>,
    #[serde(default)]
    pub locale: Option<String>,
    #[serde(default)]
    pub app_identifier: Option<String>,
    #[serde(default = "HistoryQuery::default_limit")]
    pub limit: usize,
    #[serde(default)]
    pub offset: usize,
}

impl HistoryQuery {
    fn default_limit() -> usize {
        20
    }
}

/// History entry returned to callers.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct HistoryEntry {
    pub session_id: String,
    pub started_at_ms: i64,
    pub completed_at_ms: i64,
    pub duration_ms: i64,
    #[serde(default)]
    pub locale: Option<String>,
    #[serde(default)]
    pub app_identifier: Option<String>,
    #[serde(default)]
    pub app_version: Option<String>,
    #[serde(default)]
    pub confidence_score: Option<f32>,
    pub raw_transcript: String,
    pub polished_transcript: String,
    pub preview: String,
    #[serde(default)]
    pub accuracy_flag: AccuracyFlag,
    #[serde(default)]
    pub accuracy_remarks: Option<String>,
    #[serde(default)]
    pub post_actions: Vec<HistoryPostAction>,
    #[serde(default)]
    pub metadata: serde_json::Value,
}

impl HistoryEntry {
    pub fn from_snapshot(snapshot: SessionSnapshot, accuracy: AccuracyFlag) -> Self {
        let preview = snapshot.preview();
        let SessionSnapshot {
            session_id,
            started_at_ms,
            completed_at_ms,
            locale,
            app_identifier,
            app_version,
            confidence_score,
            raw_transcript,
            polished_transcript,
            metadata,
            post_actions,
            ..
        } = snapshot;
        let duration_ms = (completed_at_ms - started_at_ms).max(0);
        Self {
            preview,
            accuracy_flag: accuracy,
            accuracy_remarks: None,
            post_actions,
            metadata,
            session_id,
            started_at_ms,
            completed_at_ms,
            duration_ms,
            locale,
            app_identifier,
            app_version,
            confidence_score,
            raw_transcript,
            polished_transcript,
        }
    }
}

/// Paginated result returned to UI/IPC clients.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct HistoryPage {
    pub entries: Vec<HistoryEntry>,
    pub next_offset: Option<usize>,
    pub total: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AccuracyUpdate {
    pub session_id: String,
    pub flag: AccuracyFlag,
    #[serde(default)]
    pub remarks: Option<String>,
}
